hypershift-operator/controllers/scheduler/autoscaler_test.go

```diff
@@ -511,3 +513,323 @@ func withHCSizeLabel(size string) func(*hyperv1.HostedCluster) {
+	largeNodes := resource.MustParse("1")    // 1 per node
+	bufferSize := resource.MustParse("1")
+
+	cfg := &schedulingv1alpha1.ClusterSizingConfiguration{
+		Spec: schedulingv1alpha1.ClusterSizingConfigurationSpec{
+			Sizes: []schedulingv1alpha1.SizeConfiguration{
+				{
+					Name: "small",
+					Management: &schedulingv1alpha1.Management{
+						NonRequestServingNodesPerZone: &smallNodes,
+					},
+				},
+				{
+					Name: "medium",
+					Management: &schedulingv1alpha1.Management{
+						NonRequestServingNodesPerZone: &mediumNodes,
+					},
+				},
+				{
+					Name: "large",
+					Management: &schedulingv1alpha1.Management{
+						Placeholders:                  0,
+						NonRequestServingNodesPerZone: &largeNodes,
+					},
+				},
+			},
+			NonRequestServingNodesBufferPerZone: &bufferSize,
+		},
+		Status: schedulingv1alpha1.ClusterSizingConfigurationStatus{
+			Conditions: []metav1.Condition{
+				{
+					Type:   schedulingv1alpha1.ClusterSizingConfigurationValidType,
+					Status: metav1.ConditionTrue,
+				},
+			},
+		},
+	}
+
+	tests := []struct {
+		name           string
+		hostedClusters []hyperv1.HostedCluster
+		machineSets    []machinev1beta1.MachineSet
+		expect         []machineSetReplicas
+	}{
+		{
+			name:        "No hosted clusters",
+			machineSets: []machinev1beta1.MachineSet{ms(1, 1), ms(2, 1), ms(3, 1)},
+			expect:      nil, // no changes
+		},
+		{
+			name:        "No hosted clusters, one machineset scaled up",
+			machineSets: []machinev1beta1.MachineSet{ms(1, 1), ms(2, 2), ms(3, 1)},
+			expect:      []machineSetReplicas{{ms(2, 2), 1}},
+		},
+		{
+			name:           "Small hosted clusters",
+			hostedClusters: []hyperv1.HostedCluster{hc(1, "small"), hc(2, "small"), hc(3, "small")}, // 0.6 should require 1 node + 1 buffer
+			machineSets:    []machinev1beta1.MachineSet{ms(1, 1), ms(2, 1), ms(3, 1)},
+			expect:         []machineSetReplicas{{ms(1, 1), 2}, {ms(2, 1), 2}, {ms(3, 1), 2}},
+		},
+		{
+			name:           "Small and medium hosted clusters, one machineset scaled up",
+			hostedClusters: []hyperv1.HostedCluster{hc(1, "medium"), hc(2, "medium"), hc(3, "small")}, // 1.2 should require 2 nodes + 1 buffer
+			machineSets:    []machinev1beta1.MachineSet{ms(1, 1), ms(2, 1), ms(3, 3)},
+			expect:         []machineSetReplicas{{ms(1, 1), 3}, {ms(2, 1), 3}},
+		},
+		{
+			name:           "Large hosted clusters, more than max",
+			hostedClusters: hcs(20, "large"), // should require 20 nodes + 1 buffer, but we're limited to 10
+			machineSets:    []machinev1beta1.MachineSet{ms(1, 1), ms(2, 1), ms(3, 1)},
+			expect:         []machineSetReplicas{{ms(1, 1), 10}, {ms(2, 1), 10}, {ms(3, 1), 10}},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			g := NewGomegaWithT(t)
+			actual := nonRequestServingMachineSetsToScale(context.Background(), cfg, test.hostedClusters, test.machineSets)
+			g.Expect(actual).To(BeEquivalentTo(test.expect))
```

Comment:
If you need `gomega`, try [this](https://github.com/openshift/rosa/blob/0f7d40978b06dbbe910a2717f167b360cc02957e/pkg/test/matchers/matchers.go#L9-L17) out, used like [this](https://github.com/openshift/rosa/blob/0f7d40978b06dbbe910a2717f167b360cc02957e/cmd/create/cluster/spec_test.go#L53) - the output of `cmp.Diff()` on large object literals is infinitely nicer than `gomega`.